### PR TITLE
fcase should save the value0's attributes

### LINKED ISSUE #4378: fcase() may trigger segfaults in certain cases
I'm using data.table 1.12.9, the most recent development version

# Minimal reproducible example

```r
qtty_row_table <- 1000000L

a <- data.table::data.table(
  V1 = rnorm(qtty_row_table),
  V2 = rnorm(qtty_row_table),
  V3 = rnorm(qtty_row_table)
)

a[, v0 := data.table::fcase(
  V1 > 0 & V2 <= 1 & V3 > 1, V2 * 100L,
  V1 > 1 & V2 <= 0 & V3 > 0, V3 * 100L,
  V1 > -1 & V2 <= 2 & V3 > 1, V1 * 100L,
  V1 > 1 & V2 <= 0 & V3 > 2, 300,
  V1 > 0 & V2 <= 1 & V3 > 1, 100,
  V1 > -1 & V2 <= 0 & V3 > -1, V1 * 100L,
  default = 0
)]
```

# Output of sessionInfo()

```
R version 3.5.2 (2018-12-20)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.5.2 tools_3.5.2   
```

### PR DESCRIPTION
**Pull request by shrektan**
Closes #4378

### The reproducible example on my Mac

(Took me some time to find this example)

```r
library(data.table)
n = 1e7
x = structure(rnorm(n), class = 'abc')
invisible(
  fcase(x <= -100, structure(x * 1.1, class = 'abc'),
        x <= -10, structure(x * 1.1, class = 'abc'),
        x <=  0,  structure(x * 1.1, class = 'abc'),
        x <=  100, structure(x * 1.1, class = 'abc'), 
        x <=  1000, structure(x * 1.1, class = 'abc'),
        x >=  1000, structure(x * 1.1, class = 'abc'))
)
```

- Before the PR: it will report an error that the "# 6" has different class
- After the PR: works fine

The cause is, the first value (assigned to variable `value0`) will be touched later,  to ensure all the values have the same class and factor levels. However, at that time, `value0` may have already been garbaged. 

The difficulty to reproduce this issue is due to the fact that **as long as the garbaged memory is not reused, it won't affect anything**. Windows and Mac should have different memory management logic and that may be the reason of the example in #4378 is only reproducible on Windows.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
I think you changed the line endings?

**Comment by jangorecki** [comment]
AFAIU reprotect with index was intentionally used to reduce overhead related to stack of protected object. Why do we need to care values that had been computed and used, are they needed anywhere downstream? Even if that fixes the problem it would be good to narrow down the issue, maybe the bug is somewhere else?

**Comment by shrektan** [comment]
You are right. Usually, the cases are very limited. However, by claiming the unused values it could be more memory efficiently.

I'll take a closer look tomorrow. The cause should be somewhere accidentally used the unprotected values.

**Comment by jangorecki** [comment]
we could as well manually unprotect at the end of each iteration, but that doesn't actually narrows down where the issue is

**Comment by shrektan** [comment]
@jangorecki  I think the cause now has been found out. I've updated my [comment](https://github.com/Rdatatable/data.table/pull/4401#issue-409567710), with a reproducible example that should work on both Windows and Mac (haven't tested it on Linux yet).

File: src/fifelse.c [hunk 1]
```diff
@@ -177,7 +179,11 @@ SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
       len0 = xlength(cons);
       len2 = len0;
       type0 = TYPEOF(outs);
-      value0 = outs;
+      v0class = PROTECT(getAttrib(outs,R_ClassSymbol)); nprotect++;
```
Comment (jangorecki) [inline src/fifelse.c:182]: couldn't we unprotect that some time sooner than at the very end of processing?

Comment (shrektan) [reply src/fifelse.c:182]: I think it doesn’t matter considering the class and the levels are very small generally. In addition, we have to keep them to the end of the loop.

**Comment by mattdowle** [comment]
I was just resolving the conflicts, and then saw the base difference in fifelse.c. The rework of fifelse.c in #4629 for 1.13.0 very likely fixed the root cause here already. For example, point 2 of that PR mentions that the first value is being referenced from within the loop but it wasn't protected, similar to the fix in this PR. I'll check the new example fails reliably under strict torture in 1.12.8, but passes strict torture in 1.13.0. If so, it's just a case of adding the test and noting it was fixed in 1.13.0.

**Comment by shrektan** [comment]
Ha, that's ok. As the issue has already been fixed, I'm closing this.

**Comment by mattdowle** [comment]
Just to double-check, with 1.12.9 (dev) as of May 21st, in Rdevel-strict-gcc with gctorture on, your example fails straight away as expected : 
```
$ Rdevel-strict-gcc
R Under development (unstable) (2020-07-14 r78854) -- "Unsuffered Consequences"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
> require(data.table)
Loading required package: data.table
data.table 1.12.9 IN DEVELOPMENT built 2020-11-03 06:43:30 UTC; mdowle using 1 threads (see ?getDTthreads).  Latest news: r-datatable.com
**********
This installation of data.table has not detected OpenMP support. It should still work but in single-threaded mode. If this is a Mac, please ensure you are using R>=3.4.0 and have followed our Mac instructions here: https://github.com/Rdatatable/data.table/wiki/Installation. This warning message should not occur on Windows or Linux. If it does, please file a GitHub issue.
**********
> n = 1e7
> x=structure(rnorm(n),class='abc')
> gctorture2(step=100)
[1] 0
> fcase(x <= -100, structure(x * 1.0, class = 'abc'),
+   x <= -10, structure(x * 1.0, class = 'abc'),
 +   x <=  0,  structure(x * 1.0, class = 'abc'),
+   x <=  100, structure(x * 1.0, class = 'abc'),
+   x <=  1000, structure(x * 1.0, class = 'abc'),
+   x >=  1000, structure(x * 1.0, class = 'abc'))
Error in fcase(x <= -100, structure(x * 1, class = "abc"), x <= -10, structure(x *  : 
  unprotected object (0x62500a0ae5e8) encountered (was REALSXP)
```

as does the original from #4378, starting with a fresh R session again:

```
$ Rdevel-strict-gcc
R Under development (unstable) (2020-07-14 r78854) -- "Unsuffered Consequences"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
> qtty_row_table <- 1000000L
> a <- data.table::data.table(
+   V1 = rnorm(qtty_row_table),
+   V2 = rnorm(qtty_row_table),
+   V3 = rnorm(qtty_row_table)
+ )
> gctorture2(step=100)
[1] 0
> a[, v0 := data.table::fcase(
+   V1 > 0 & V2 <= 1 & V3 > 1, V2 * 100L,
+   V1 > 1 & V2 <= 0 & V3 > 0, V3 * 100L,
+   V1 > -1 & V2 <= 2 & V3 > 1, V1 * 100L,
 +   V1 > 1 & V2 <= 0 & V3 > 2, 300,
+   V1 > 0 & V2 <= 1 & V3 > 1, 100,
+   V1 > -1 & V2 <= 0 & V3 > -1, V1 * 100L,
+   default = 0
+ )]
=================================================================
==2626609==ERROR: AddressSanitizer: heap-use-after-free on address 0x7f525593b800 at pc 0x55b1f2a5c7af bp 0x7ffde762c130 sp 0x7ffde762c120
READ of size 1 at 0x7f525593b800 thread T0
    #0 0x55b1f2a5c7ae in CHK /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:173
    #1 0x55b1f2a5c7ae in TYPEOF /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:3755
    #2 0x55b1f271cfaf in Rf_getAttrib /home/mdowle/build/R-devel-strict-gcc/src/main/attrib.c:164
    #3 0x7f525a582a04 in fcaseR (/home/mdowle/build/rchk/trunk/packages/lib/data.table/libs/datatable.so+0x1e4a04)
    #4 0x55b1f2826d4a in R_doDotCall /home/mdowle/build/R-devel-strict-gcc/src/main/dotcode.c:604
    #5 0x55b1f297d818 in bcEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:7670
    #6 0x55b1f29b8787 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:723
    #7 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #8 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #9 0x55b1f29b88fb in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:846
    #10 0x55b1f29c391d in do_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3335
    #11 0x55b1f2ac8eb0 in do_internal /home/mdowle/build/R-devel-strict-gcc/src/main/names.c:1385
    #12 0x55b1f29b8b5e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:798
    #13 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #14 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #15 0x55b1f2983fc3 in bcEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:7082
    #16 0x55b1f29b8787 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:723
    #17 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #18 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #19 0x55b1f2acef61 in dispatchMethod /home/mdowle/build/R-devel-strict-gcc/src/main/objects.c:436
    #20 0x55b1f2acf6a5 in Rf_usemethod /home/mdowle/build/R-devel-strict-gcc/src/main/objects.c:476
    #21 0x55b1f29c606f in Rf_DispatchOrEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3614
    #22 0x55b1f2c55f9f in R_DispatchOrEvalSP /home/mdowle/build/R-devel-strict-gcc/src/main/subset.c:633
    #23 0x55b1f2c55f9f in do_subset /home/mdowle/build/R-devel-strict-gcc/src/main/subset.c:653
    #24 0x55b1f29b8b5e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:798
    #25 0x55b1f2a46c35 in Rf_ReplIteration /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:264
    #26 0x55b1f2a47620 in R_ReplConsole /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:314
    #27 0x55b1f2a477b6 in run_Rmainloop /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:1113
    #28 0x55b1f2a47826 in Rf_mainloop /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:1120
    #29 0x55b1f269be41 in main /home/mdowle/build/R-devel-strict-gcc/src/main/Rmain.c:29
    #30 0x7f52622db0b2 in __libc_start_main (/usr/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #31 0x55b1f269e67d in _start (/home/mdowle/build/R-devel-strict-gcc/bin/exec/R+0x7e367d)

0x7f525593b800 is located 0 bytes inside of 8000048-byte region [0x7f525593b800,0x7f52560dca30)
freed by thread T0 here:
    #0 0x7f52636e67cf in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x10d7cf)
    #1 0x55b1f2a9482e in ReleaseLargeFreeVectors /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:1114
    #2 0x55b1f2a9482e in RunGenCollect /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:1896
    #3 0x55b1f2a9482e in R_gc_internal /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:3125
    #4 0x55b1f2ab6edb in Rf_allocVector3 /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:2775
    #5 0x55b1f2b876d0 in Rf_allocVector ../../src/include/Rinlinedfuns.h:593
    #6 0x55b1f2b876d0 in numeric_relop /home/mdowle/build/R-devel-strict-gcc/src/main/relop.c:334
    #7 0x55b1f2b907bb in do_relop_dflt /home/mdowle/build/R-devel-strict-gcc/src/main/relop.c:127
    #8 0x55b1f2b98400 in do_relop /home/mdowle/build/R-devel-strict-gcc/src/main/relop.c:73
    #9 0x55b1f29b914e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:830
    #10 0x55b1f29c1cfd in Rf_evalList /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3058
    #11 0x55b1f29b8db8 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:817
    #12 0x55b1f29c1cfd in Rf_evalList /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3058
    #13 0x55b1f29b8db8 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:817
    #14 0x7f525a582892 in fcaseR (/home/mdowle/build/rchk/trunk/packages/lib/data.table/libs/datatable.so+0x1e4892)
    #15 0x55b1f2826d4a in R_doDotCall /home/mdowle/build/R-devel-strict-gcc/src/main/dotcode.c:604
    #16 0x55b1f297d818 in bcEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:7670
    #17 0x55b1f29b8787 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:723
    #18 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #19 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #20 0x55b1f29b88fb in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:846
    #21 0x55b1f29c391d in do_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3335
    #22 0x55b1f2ac8eb0 in do_internal /home/mdowle/build/R-devel-strict-gcc/src/main/names.c:1385
    #23 0x55b1f29b8b5e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:798
    #24 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #25 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #26 0x55b1f2983fc3 in bcEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:7082
    #27 0x55b1f29b8787 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:723
    #28 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #29 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #30 0x55b1f2acef61 in dispatchMethod /home/mdowle/build/R-devel-strict-gcc/src/main/objects.c:436
    #31 0x55b1f2acf6a5 in Rf_usemethod /home/mdowle/build/R-devel-strict-gcc/src/main/objects.c:476
    #32 0x55b1f29c606f in Rf_DispatchOrEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3614

previously allocated by thread T0 here:
    #0 0x7f52636e6bc8 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x10dbc8)
    #1 0x55b1f2ab808b in Rf_allocVector3 /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:2806
    #2 0x55b1f26ea9ca in Rf_allocVector ../../src/include/Rinlinedfuns.h:593
    #3 0x55b1f26ea9ca in R_allocOrReuseVector /home/mdowle/build/R-devel-strict-gcc/src/main/arithmetic.h:95
    #4 0x55b1f26ea9ca in real_binary /home/mdowle/build/R-devel-strict-gcc/src/main/arithmetic.c:983
    #5 0x55b1f26ea9ca in R_binary /home/mdowle/build/R-devel-strict-gcc/src/main/arithmetic.c:691
    #6 0x55b1f26fb76c in do_arith /home/mdowle/build/R-devel-strict-gcc/src/main/arithmetic.c:520
    #7 0x55b1f29b914e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:830
    #8 0x7f525a582916 in fcaseR (/home/mdowle/build/rchk/trunk/packages/lib/data.table/libs/datatable.so+0x1e4916)
    #9 0x55b1f2826d4a in R_doDotCall /home/mdowle/build/R-devel-strict-gcc/src/main/dotcode.c:604
    #10 0x55b1f297d818 in bcEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:7670
    #11 0x55b1f29b8787 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:723
    #12 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #13 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #14 0x55b1f29b88fb in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:846
    #15 0x55b1f29c391d in do_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3335
    #16 0x55b1f2ac8eb0 in do_internal /home/mdowle/build/R-devel-strict-gcc/src/main/names.c:1385
    #17 0x55b1f29b8b5e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:798
    #18 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #19 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #20 0x55b1f2983fc3 in bcEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:7082
    #21 0x55b1f29b8787 in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:723
    #22 0x55b1f29bc234 in R_execClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1888
    #23 0x55b1f29bde61 in Rf_applyClosure /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:1814
    #24 0x55b1f2acef61 in dispatchMethod /home/mdowle/build/R-devel-strict-gcc/src/main/objects.c:436
    #25 0x55b1f2acf6a5 in Rf_usemethod /home/mdowle/build/R-devel-strict-gcc/src/main/objects.c:476
    #26 0x55b1f29c606f in Rf_DispatchOrEval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:3614
    #27 0x55b1f2c55f9f in R_DispatchOrEvalSP /home/mdowle/build/R-devel-strict-gcc/src/main/subset.c:633
    #28 0x55b1f2c55f9f in do_subset /home/mdowle/build/R-devel-strict-gcc/src/main/subset.c:653
    #29 0x55b1f29b8b5e in Rf_eval /home/mdowle/build/R-devel-strict-gcc/src/main/eval.c:798
    #30 0x55b1f2a46c35 in Rf_ReplIteration /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:264
    #31 0x55b1f2a47620 in R_ReplConsole /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:314
    #32 0x55b1f2a477b6 in run_Rmainloop /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:1113
    #33 0x55b1f2a47826 in Rf_mainloop /home/mdowle/build/R-devel-strict-gcc/src/main/main.c:1120

SUMMARY: AddressSanitizer: heap-use-after-free /home/mdowle/build/R-devel-strict-gcc/src/main/memory.c:173 in CHK
Shadow bytes around the buggy address:
  0x0feacab1f6b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feacab1f6c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feacab1f6d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feacab1f6e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0feacab1f6f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0feacab1f700:[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0feacab1f710: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0feacab1f720: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0feacab1f730: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0feacab1f740: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0feacab1f750: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==2626609==ABORTING
```

But with 1.13.0 both examples complete cleanly under torture:

```
$ Rdevel-strict-gcc
R Under development (unstable) (2020-07-14 r78854) -- "Unsuffered Consequences"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)
> require(data.table)
Loading required package: data.table
data.table 1.13.0 using 1 threads (see ?getDTthreads).  Latest news: r-datatable.com
**********
This installation of data.table has not detected OpenMP support. It should still work but in single-threaded mode.
This is Linux. This warning should not normally occur on Windows or Linux where OpenMP is turned on by data.table's configure script by passing -fopenmp to the compiler. If you see this warning on Windows or Linux, please file a GitHub issue.
**********
> n = 1e7
> x=structure(rnorm(n),class='abc')
> gctorture2(step=100)
[1] 0
> ans = fcase(x <= -100, structure(x * 1.0, class = 'abc'),
+   x <= -10, structure(x * 1.0, class = 'abc'),
+   x <=  0,  structure(x * 1.0, class = 'abc'),
 +   x <=  100, structure(x * 1.0, class = 'abc'),
 +   x <=  1000, structure(x * 1.0, class = 'abc'),
 +   x >=  1000, structure(x * 1.0, class = 'abc'))
> head(ans)
[1]  1.55888176  0.63787126  0.36124433 -0.60787011  1.13407518  0.01810095
> qtty_row_table <- 1000000L
> a <- data.table::data.table(
+   V1 = rnorm(qtty_row_table),
+   V2 = rnorm(qtty_row_table),
+   V3 = rnorm(qtty_row_table)
+ )
> a[, v0 := data.table::fcase(
+   V1 > 0 & V2 <= 1 & V3 > 1, V2 * 100L,
+   V1 > 1 & V2 <= 0 & V3 > 0, V3 * 100L,
+   V1 > -1 & V2 <= 2 & V3 > 1, V1 * 100L,
+   V1 > 1 & V2 <= 0 & V3 > 2, 300,
+   V1 > 0 & V2 <= 1 & V3 > 1, 100,
+   V1 > -1 & V2 <= 0 & V3 > -1, V1 * 100L,
+   default = 0
+ )]
> a
                  V1          V2           V3         v0
      1:  0.15551785  0.51119430  1.714512282  51.119430
      2: -0.60809308  0.90497410  0.157391528   0.000000
      3:  0.23065675  1.24415562  0.064219203   0.000000
      4:  1.22198746 -0.13954674 -1.746241582   0.000000
      5:  1.34503749  0.01610219 -0.005848846   0.000000
     ---                                                
 999996: -0.44473961 -1.22216216 -0.633922180 -44.473961
 999997:  0.23423232 -0.06269643  0.643722770  23.423232
 999998: -1.78482175 -0.07132834  0.473451821   0.000000
 999999:  0.03919514 -0.59569046  0.366203037   3.919514
1000000:  0.34639662  1.13805966 -0.198578115   0.000000
>
```

**Pull request closed without merging**